src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,222 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
+// it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <uv.h>
+
+#include "nvim/api/private/helpers.h"
+#include "nvim/api/private/defs.h"
+#include "nvim/lua/executor.h"
+#include "nvim/vim.h"
+#include "nvim/version.h"
+
+#ifdef __linux__
+
+#define XLIB_NAME "libX11.so.6"
+typedef void *_XDisplay;
+typedef struct {
+    char data[18];
+} _XkbStateRec;
+typedef int _XStatus;
+
+enum { kLockedModsOffset = 9 };
+enum { _XkbUseCoreKbd = 0x0100 };
+#endif
+
+#ifdef WIN32
+#include <Winuser.h>
+#endif
+
+#ifdef __APPLE__
+#define Boolean Boolean_I_Dont_Care
+#include <IOKit/IOKitLib.h>
+#include <IOKit/IOReturn.h>
+#include <IOKit/hidsystem/IOHIDLib.h>
+#include <IOKit/hidsystem/IOHIDParameter.h>
+#include <CoreFoundation/CoreFoundation.h>
+#undef Boolean
+#endif
+
+enum {
+    kScrollLock = 0x80,
+    kNumLock = 0x10,
+    kCapsLock = 0x02,
+};
+
+
+#if defined(__linux__)
+static bool xlib_opened = false;
+static uv_lib_t xlib;
+
+/// A helper function to be used to load the libX11
+static bool os_xlib_open(const char **const err)
+{
+  if (xlib_opened) {
+    return true;
+  }
+  const int udl_ret = uv_dlopen(XLIB_NAME, &xlib);
+  if (udl_ret != 0) {
+    *err = "Failed to open library libX11";
+  } else {
+    xlib_opened = true;
+  }
+  return xlib_opened;
+}
+
+/// A helper function to be used to get the pointer of the function to be used
+static bool os_xlib_dlsym(const char *const name, void **const ptr,
+                       const char **const err)
+{
+  const int uds_ret = uv_dlsym(&xlib, name, ptr);
+  if (uds_ret != 0) {
+    *err = "Failed to retrieve the function from X11 library";
+    return false;
+  }
+  return true;
+}
+
+/// Get the status of the Capslock, Numlock and Scrolllock in linux
+static bool os_get_locks_status(char *const mods, const char **const err)
+{
+#define CLOSE_DISPLAY(display) \
+  do { \
+    const int e = _XCloseDisplay(display); \
+    if (e) { \
+      *err = "XCloseDisplay unable to close the display"; \
+      return false; \
+    } \
+  } while (0)
+  *mods = 0;
+  *err = NULL;
+  if (!os_xlib_open(err)) {
+    return false;
+  }
+  _XDisplay *(*_XOpenDisplay)(void *);
+  int (*_XCloseDisplay)(_XDisplay *);
+  _XStatus (*_XkbGetState)(_XDisplay *, unsigned, _XkbStateRec *);
+  if (!os_xlib_dlsym("XOpenDisplay", (void **)&_XOpenDisplay, err)
+      || !os_xlib_dlsym("XCloseDisplay", (void **)&_XCloseDisplay, err)
+      || !os_xlib_dlsym("XkbGetState", (void **)&_XkbGetState, err)) {
+    return false;
+  }
```

Comment:
This all should really go to `os_xlib_init` (renamed from `os_xlib_open`): there is no need to reretrieve symbols each time. Do not know why I did not think of that earlier.

Reply:
Oks, no problem